tools/switch.sh

```diff
@@ -0,0 +1,211 @@
+  ./switch.sh fabric | grep '^PAX ID:' -A4                                      # display the number of endpoints attached to each PAX
+  ./switch.sh fabric | grep -e "Physical Port ID " -A2 -e "^PAX ID:"            # display the drive physical ports along with their PDFIDs
+
+  ./switch.sh fabric gfms-dump                                                  # dump GFMS database (same as ./switch.sh fabric)
+  ./switch.sh fabric gfms-dump | grep -e "Execute" -e "^PAX ID:"                # display PAX ID associated with each /dev/switchtec device
+  ./switch.sh fabric gfms-dump | grep "Function 0 (SRIOV-PF)" -A1 | grep PDFID  # display the list of physical device fabric IDs for the drives attached
+
+  ./switch.sh fabric topo-info                                                  # show topology info including PCIe rates
+
+  ./switch.sh cmd fw-info                                                       # return information on the currently flashed firmware
+  ./switch.sh cmd fw-info | grep -A1 'Currently Running'                        # display currently running firmware
+EOF
+}
+
+# execute <fn<path>> <args...>
+execute() {
+    local FUNCTION=$1 ARGS=( "${@:2}" )
+
+    if [ "$(type -t "$FUNCTION")" != "function" ]; then
+        echo "$1 is not a function."
+        exit 1
+    fi
+
+    local SWITCHES=("/dev/switchtec0" "/dev/switchtec1")
+    for SWITCH in "${SWITCHES[@]}";
+    do
+        "$FUNCTION" "$SWITCH" "${ARGS[@]}"
+    done
+}
+
+displaySlotStatus() {
+    local SWITCH_NAME=$1
+    local PAX_ID
+
+    # Physical slot ids are set into the hardware. These are the mappings
+    declare -a PAX0_SlotFromPhysicalPort=(
+        [8]=8
+        [10]=7
+        [12]=15
+        [14]=16
+        [16]=17
+        [18]=18
+        [20]=14
+        [22]=13
+        [48]=12
+    )
+    declare -a PAX1_SlotFromPhysicalPort=(
+        [8]=4
+        [10]=5
+        [12]=6
+        [14]=2
+        [16]=1
+        [18]=9
+        [20]=10
+        [22]=11
+        [48]=3
+    )
+
+    # Make sure we can get the PAX ID
+    if [ ! "$(switchtec fabric gfms-dump "$SWITCH_NAME" | grep "^PAX ID:" | awk '{print $3}')" ]; then
+        exit $?
+    fi
+
+    PAX_ID=$(switchtec fabric gfms-dump "$SWITCH_NAME" | grep "^PAX ID:" | awk '{print $3}')
+
+    if ! [[ $PAX_ID =~ [[:digit:]]+ ]]; then
+        echo "$PAX_ID is not an integer"
+        exit 1
+    elif ! (( PAX_ID >= 0 && PAX_ID <= 1 )); then
+        echo "$PAX_ID not in range 0-1"
+        exit 1
+    fi
+
+    # Grab the attach status of each physical port
+    mapfile -t physicalPortStatus < <(switchtec fabric gfms-dump "$SWITCH" | grep " Physical Port ID")
+
+    local physicalPortString
+    for physicalPortString in "${physicalPortStatus[@]}";
+    do
+        local PHYSICAL_PORT_ID
+        PHYSICAL_PORT_ID=$(echo "$physicalPortString" | awk '{print $4}')
+        case $PAX_ID in
```

Comment:
I wonder if you could declare an "array of arrays" above, near line 106, then index that value here. That way you're not mixing data with code.

`declare -a PAXPORTS=("PAX0_SlotFromPhysicalPort", "PAX1_SlotFromPhysicalPort")`

```
PAX="${PAXPORTS[$PAX_ID]}" # I'm not sure if this is correct
SLOT=${PAX[$PHYSICAL_PORT_ID]}
```

Just a thought, probably not worth the effort

Reply:
The reading I did on bash arrays said multi-dimensional arrays are not supported. Python was usually suggested.
I thought about keeping it in a single array and using the pax ID as the multiplier to move to the correct section, but I thought this was clearer and easier to verify against nnf-ec/pkg/manager-fabric/config_default.yaml